### ISSUE TITLE
Information/Error? None of the codes provided are in the value set, although the code is included

### ISSUE THREAD

**Issue by ziegm** [hl7ch/ch-core#27]
Path:
PractitionerRole/HPWengerRole: PractitionerRole.code[0]
```
<code>
       <coding>
            <system value="http://snomed.info/sct"/>
            <code value="309343006"/>
            <display value="Physician"/>
       </coding>
</code>
```

Severity:
information

**Message:
None of the codes provided are in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/HCProfessional.hcProfession (http://fhir.ch/ig/ch-epr-term/ValueSet/HCProfessional.hcProfession, and a code is recommended to come from this value set) (codes = http://snomed.info/sct#309343006)**

But the Code 309343006 is included in the ValueSet:
http://build.fhir.org/ig/hl7ch/ch-epr-term/branches/master/ValueSet-HCProfessional.hcProfession.html

**Comment by ziegm** [comment]
Further examples with the same problem:

C:\Users\micha\Documents\GitHub\ch-core\examples\documentreference\Docu.xml

- **DocumentReference.type** -> None of the codes provided are in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.typeCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.typeCode, and a code is recommended to come from this value set) (codes = http://snomed.info/sct#371532007)
- **DocumentReference.category[0]** -> None of the codes provided are in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.classCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.classCode, and a code is recommended to come from this value set) (codes = http://snomed.info/sct#371525003)
- **DocumentReference.securityLabel[0]** -> None of the codes provided are in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.confidentialityCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.confidentialityCode, and a code is recommended to come from this value set) (codes = http://snomed.info/sct#17621005)
- **DocumentReference.content[0].attachment.contentType** -> The value provided ('**application/pdf**') is not in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.mimeType (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.mimeType, and a code is recommended to come from this value set) (error message = Unknown code system http://terminology.hl7.org/CodeSystem/v3-mediaType) #19 
- **DocumentReference.content[0].attachment.language** -> The value provided ('**de-CH**') is not in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.languageCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.languageCode, and a code is recommended to come from this value set) (error message = Unknown code system urn:oid:2.16.840.1.113883.6.316)
- **DocumentReference.content[0].format** -> The Coding provided is not in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.formatCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.formatCode, and a code is recommended to come from this value set) (error message = Unknown code system urn:oid:2.16.756.5.30.1.127.3.10.10)

**Comment by ziegm** [comment]
**information** if the binding of the value set is **preferred**
**error** if the binding of the value set is **required**

**Comment by oliveregger** [comment]
Validating Code in ValueSet: if Codesystem is not complete (e.g: SNOMED  CT code), the validation of the code will fail ( Unable to resolve system "+c.getSystem()+" - system is not complete), ValueSetCheckerSimple [line: 75] - validateCode(CodeableConcept),  if this check is excluded, validation will fail because if code in CodeSystem is one more time checked 

If the code cannot be validated locally (see above) the terminology server is asked to validate the ValueSet, however our swiss specific ValueSets are not known to tx.fhir.org.

Different options:

- update the validator to validate the ValueSet locally
- ValueSet definition needs to be changed (maybe providing an expansion, but then the designations get lost in creating the narrative)
- own terminology server with EPR ValueSets provided or tx.fhir.org 

Next step:
- Need to build a sample within the ch-epr-term to illustrate the problem without the dependencies issues with ch-core

**Comment by ziegm** [comment]
Build a sample within the ch-epr-term (branch testing):
https://github.com/hl7ch/ch-epr-term/tree/testing

StructureDefinition:
https://github.com/hl7ch/ch-epr-term/blob/testing/resources/structuredefinition/ch-epr-term-documentreference.xml
Example:
https://github.com/hl7ch/ch-epr-term/blob/testing/examples/documentreference/Docu.xml

**Comment by oliveregger** [comment]
https://chat.fhir.org/#narrow/stream/179177-conformance/topic/Validator.20failure.20for.20own.20ValueSet.20defintion.20with.20SNOMED.20CT

We define ValueSets which contain a finite list of SNOMED CT codes and and additional codes from an other CodeSystem. We defined them via compose even though it is a list of fixed codes because we have designations for the different languages. We bind this ValueSet to a profiled DocumentReference (see example).

The Java Validator/IG Publisher fails for this examples with the message:

None of the codes provided are in the value set http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.typeCode (http://fhir.ch/ig/ch-epr-term/ValueSet/DocumentEntry.typeCode, and a code from this value set is required) (codes = http://snomed.info/sct#371532007)
A first analysis revealed that the Validator does not validate the code locally with the provided ValueSet because the code is from SNOMED CT which is not a complete CodeSystem. In the next step the Validator then sends the ValueSet together with the code to tx.fhir.org. This fails now because the second code system is not known to tx.fhir.org (it is defined currently only in our implementation guide).

This leads to three questions I would happy to get some feedback:
- is our approach correct with defining ValueSets with compose or should we provide already expanded ValueSets in our Implementation Guide?
- could the validation logic be extend, that if a code can be validated within a ValueSet event though it is not a complete CodeSystem so that the code does not need to be validated with tx.fhir.org? (This would not work in all cases would be more an optimization)? if yes, we would start to work on a pull request
- can also a local code system be supplied to tx.fhir.org (inlined in the ValueSet)?

**Comment by oliveregger** [comment]
1) approach is correct
2) validation logic has been added by grahame
3) local code system is not necessary, e.g. currently not finished approach for switzerland, local snomed code system will be published in an snomed ct edition next year

https://chat.fhir.org/#narrow/stream/179177-conformance/topic/Validator.20failure.20for.20own.20ValueSet.20defintion.20with.20SNOMED.20CT

fixed with ig publisher v1.0.18-SNAPSHOT
